java-scala-diff.md

```diff
@@ -0,0 +1,285 @@
+def nickname = firstName
+```
+
+#### Type parameters
+```java
+// in Java, usually named T or another single upper-case letter
+void <T, K, V> doSomething(List<T> list, Map<K, V> map) { ... }
+```
+```scala
+// in Scala, usually first letters of alphabet or words in UpperCamelCase
+def doSomething[A, Key, Value](list: List[A], map: Map[Key, Value]): Unit = ...
+```
+
+#### Curly braces
+```java
+// in Java, surrounding spaces are optional
+a = new int[] {5, 6};
+b = new int[] { 5, 6 };
+```
+```scala
+// in Scala, separate by spaces
+tuples.filter { case (s1, s2) => s1 == s2 }
+```
+```java
+// in Java, opening brace may be on next line
+void foo()
+{
+  ...
+}
+```
+```scala
+// in Scala, opening brace on same line, otherwise compiler might infer semicolon unexpectedly
+def foo(): Unit = {
+  ...
+}
+```
+```java
+// in Java, always use with if/else/for/do/while
+if (foo) {
+  doSomething();
+} else {
+  doSomethingElse();
+}
+```
+```scala
+// in Scala, omit if a pure-functional operation and all branches are single-line
+val x = if (foo)
+  doSomething()
+else
+  doSomethingElse()
+
+// don’t omit if no “else”
+if (foo) {
+  doSomething()
+}
+
+// always omit in “case”
+val x = foo match {
+  case 1 =>
+    doSomething()
+    doSomethingMore()
+  case 2 => doSomethingElse()
+}
+
+// omit if “yield”
+for (i <- 1 to 5) yield i
+
+// don’t omit if no “yield”
+for (i <- 1 to 5) {
+  println(i)
+}
+
+// omit in one-line methods
+def add(a: Int, b: Int): Int = a + b
+
+// OK to put it on the next line
+def sum(ls: List[String]): Int =
+  ls.map(_.toInt).foldLeft(0)(_ + _)
+
+// omit in methods consisting of a single “match”
+def sum(ls: List[Int]): Int = ls match {
+  case hd :: tail => hd + sum(tail)
+  case Nil => 0
+}
+
+```
+
+#### Ternary operator
+```java
+// in Java
+x = foo ? bar : baz;
+```
+```scala
+// in Scala
+x = if (foo) bar else baz
+```
+
+#### Class contents order
+In Java, “some logical order”. In Scala, fields before methods.
+
+#### Annotations
+```java
+// in Java, may be on the same line
+@Override public int hashCode() {
+  ...
+}
+
+@Partial @Mock DataLoader loader;
+```
+```scala
+// in Scala, each annotatin on its own line
+@transaction
+@throws(classOf[IOException])
+override protected final def foo(): Unit = {
+  ...
+}
+```
+
+#### Modifier order
+```java
+// in Java
+public protected private abstract default static final transient volatile synchronized native strictfp
+```
+```scala
+// in Scala
+override protected private implicit final lazy
+```
+
+#### Javadoc, Scaladoc
+```java
+// Javadoc
+
+/**
+ * Does something.
+ *
+ * Some more details.
+ */
+void doSomething() {
+  ...
+}
+```
+```scala
+// Scaladoc with asterisks in second column
+
+/** Does something.
+ *
+ *  Some more details.
+ */
+def doSomething(): Unit = {
+  ...
+}
+
+// Scaladoc with asterisks in third column
+
+/** Does something.
+  *
+  * Some more details.
+  */
+def doSomething(): Unit = {
```

Comment:
Is this saying we support either of these styles?
Should we make a recommendation for one over the other?

Reply:
Done. 🤷‍♀️

Reply:
😛 

Seems like we'd want to pick one so we have consistency. Do you see one style more prevalent than the other? If so, let's go with that one and recommend not using the other one.

Reply:
Done